File: python/ecuaciones_no_lineales/codigos/biseccion.py
import math
import matplotlib.pyplot
from sympy import *


def biseccion(f, a, b, tol, iterMax):
    f1 = sympify(f)
    fa = f1.subs('x', a)
    fb = f1.subs('x', b)
    if (fa * fb > 0):
        x = []
        k = []
        error = []
        print('No cumple la condición del Teorema de Bolzano.')
    else:
        error = tol + 1
        k = 0
        it = []
        e = []
        while (tol < error and k < iterMax):
            k = k + 1
            x = (a + b) / 2
            fa = f1.subs('x', a)
            fx = f1.subs('x', x)
            error = abs(fx)
            it.append(k)
            e.append(error)
            if (fa * fx < 0):
                b = x
            else:
                a = x
    matplotlib.pyplot.plot(it, e)
    ans = {
        "x": x,
        "k": k,
        "error": error
    }
    return ans

# biseccion('exp(x)-x-2',0,2,10**-8,10)

File: octave/ecuaciones_no_lineales/biseccion.m
function [x, error, k] = biseccion(f, a, b, tol)
  f1 = str2func(f); % 1
  % f1 = matlabFunction(sym(f)); % 2

  if (f1(a) * f1(b) <= 0)
    error = tol + 1;
    k = 0; 
    e = [];

    while (tol < error) 
      x = (a + b) / 2;

      if (f1(a) * f1(x) <= 0) 
        b = x;
      else 
        a = x;
      end  

      error = abs(f1(x));
      e = [e error];
      ++k;
    end
    
    plot(1 : k, e)
    xlabel('iter (k)')
    ylabel('Error (|f(x_k)|)')
    title('Error del Metodo de la Biseccion')
  else
    display('No cumple la condición del Teorema de Bolzano.')  
  end
end

% 1
% [x, err, iter] = biseccion('@(x)exp(x)-x-2', 0, 2, 10^-8)

% 2
% [x, err, iter] = biseccion(exp(x)-x-2, 0, 2, 10^-8)